THORN-2063: Migrate MP Fault Tolerance implementation to SmallRye

antoinesd opened the pull request.

antoinesd committed:
THORN-2063: Migrate MP Fault Tolerance implementation to SmallRye

mkouba commented on fractions/microprofile/microprofile-fault-tolerance/pom.xml:
```diff
@@ -35,6 +35,7 @@
     <swarm.fraction.tags>Eclipse MicroProfile,MicroServices,Fault Tolerance</swarm.fraction.tags>
     <version.arquillian-weld-embedded>2.0.0.Beta5</version.arquillian-weld-embedded>
     <version.awaitility>3.1.0</version.awaitility>
+    <version.smallrye-fault-tolerance>1.0.0</version.smallrye-fault-tolerance>
```
This should go into the `build-parent` pom:
https://github.com/thorntail/thorntail/blob/master/build-parent/pom.xml#L112-L116

mkouba replied:
Also add `io.smallrye:smallrye-fault-tolerance` dependency to the `<dependencyManagement>` section in build-parent; there is a `<!-- SmallRye -->` comment where other MPs reside.

mkouba commented on fractions/microprofile/microprofile-fault-tolerance/module.conf:
```diff
@@ -8,8 +8,7 @@ org.jboss.logging
 org.eclipse.microprofile.config.api
-
-org.eclipse.microprofile.faulttolerance export=true
+org.jboss.weld.core export=true
```
Why is this module dependency needed?

mkouba commented on fractions/microprofile/microprofile-fault-tolerance/src/main/java/org/wildfly/swarm/microprofile/faulttolerance/MicroProfileFaultToleranceFraction.java:
```diff
@@ -16,27 +16,18 @@
-
-    @AttributeDocumentation("Enable/disable synchronous circuit breaker functionality. If disabled, `CircuitBreaker#successThreshold()` of value greater than 1 is not supported. Moreover, circuit breaker does not necessarily transition from `CLOSED` to `OPEN` immediately when a fault tolerance operation completes. However, applications are encouraged to disable this feature on high-volume circuit ...
-    @Configurable("swarm.microprofile.fault-tolerance.synchronous-circuit-breaker")
-    private Defaultable<Boolean> synchronousCircuitBreaker = Defaultable.bool(true);
```
If we remove this it won't be possible to configure this functionality using yaml config files. Are you sure it's a good idea?

antoinesd committed:
THORN-2063: Migrate MP Fault Tolerance implementation to SmallRye
integrate feedback

mkouba commented on fractions/microprofile/microprofile-fault-tolerance/src/main/java/org/wildfly/swarm/microprofile/faulttolerance/MicroProfileFaultToleranceFraction.java:
```diff
@@ -39,4 +40,5 @@ public boolean isSynchronousCircuitBreakerEnabled() {
     @AttributeDocumentation("Enable/disable synchronous circuit breaker functionality. If disabled, `CircuitBreaker#successThreshold()` of value greater than 1 is not supported. Moreover, circuit breaker does not necessarily transition from `CLOSED` to `OPEN` immediately when a fault tolerance operation completes. However, applications are encouraged to disable this feature on high-volume circuit ...
     @Configurable("swarm.microprofile.fault-tolerance.synchronous-circuit-breaker")
     private Defaultable<Boolean> synchronousCircuitBreaker = Defaultable.bool(true);
```
Is the config option used somewhere?

antoinesd replied:
Excellent question. After checking the answer is "no" :-(. Do you have a suggestion to make it usable? I think about creating a configuration bean in smallrye ft that we would specialize here... Do you have a better idea?

mkouba replied:
Yes, you could either provide a bean which provides additional configuration or even try to implement an MP `ConfigSource`?

antoinesd replied:
I added a new ConfigSource to read config from Thorntail but it is ignored. Any clue @mkouba ?

antoinesd committed:
THORN-2063: Migrate MP Fault Tolerance implementation to SmallRye
Remove randomly failing TCK test

mkouba commented:
@antoinesd `@DeploymentModule(name = "org.wildfly.swarm.microprofile.faulttolerance", metaInf = MetaInfDisposition.IMPORT, export = true, services = Module.ServiceHandling.IMPORT)` adds the "main" slot by default (see also `org.wildfly.swarm.spi.api.annotations.DeploymentModule.slot()`). You should probably add "deployment" slot and put `ConfigurableConfigSource` under `org.wildfly.swarm.microprofile.faulttolerance.deployment` package.

antoinesd commented:
@mkouba I changed config to
`@DeploymentModule(name = "org.wildfly.swarm.microprofile.faulttolerance", metaInf = MetaInfDisposition.IMPORT, export = true, services = Module.ServiceHandling.IMPORT, slot = "deployment")`
and moved config source classes from `runtime` to `deployment` package.
Now I have a NCDFE on `org/wildfly/swarm/spi/api/config/ConfigView`. In the previous version, injection point of type `ConfigView` was unsatisfied...
[EDIT] Just pushed this new version

mkouba commented:
I think we need to do the same what `io.thorntail:cdi-config` does:

1. modify `module.conf` so that it also contains: 

```
org.wildfly.swarm.spi
org.wildfly.swarm.spi:runtime
org.wildfly.swarm.container:runtime export=true
```
("deployment" slot module does not declare a dependency on SPI module where `ConfigView` is defined)

2. Copy the functionality from `InjectConfigViewExtensionFactory` - I'm afraid that `CDI.current().select(ConfigView.class)` wouldn't work because the `ConfigView` bean lives in the internal CDI container instance

antoinesd commented:
Thanks Martin

mkouba commented:
In fact, I think we can move the config source (`ConfigView` to `ConfigSource` adapter) to the MP config fraction, in the MP config fraction declare a dependency on `cdi-config` fraction (so that we're able to inject `ConfigView`) and then it would be possible to access Thorntail internal config (yaml etc.) using MP config structures. @kenfinnigan @antoinesd @michalszynkiewicz WDYT?

antoinesd commented:
+1
That was my initial idea before I start loosing so much time in this class loader / container / dependencies hell ;-)

kenfinnigan commented:
I think that makes sense @mkouba

mkouba commented:
Ok, I've created https://issues.jboss.org/browse/THORN-2118 and will send a separate PR later today.

kenfinnigan commented:
What's still required to get this working?

antoinesd committed:
THORN-2063: Migrate MP Fault Tolerance implementation to SmallRye
Add support specific CircuitBreaker Thorntail config thru yaml

antoinesd commented:
It's working now. Yaml configuration is taken into account and has priority on MP config

mkouba commented on fractions/microprofile/microprofile-fault-tolerance/src/main/java/org/wildfly/swarm/microprofile/faulttolerance/MicroProfileFaultToleranceFraction.java:
```diff
@@ -16,27 +16,24 @@
 
     public MicroProfileFaultToleranceFraction() {
-    }
+        String synchronousCircuitBreaker = System.getProperty("swarm.microprofile.fault-tolerance.synchronous-circuit-breaker");
```
I'm not so sure this would work. `@Configurable` is repeatable, so you can try to add another `@Configurable()` alias to the config property instead. 

By the way, I've created https://github.com/smallrye/smallrye-fault-tolerance/issues/10 ...

Anyway, a simple test would be good.

mkouba committed:
THORN-2063: MP FT fraction - minor updates

- keep Fraction config attribute for docs generation
- add test
- upgrade smallrye-faulttolerance to 1.0.1
